File: backend/src/utils/attendanceHelper.js
import prisma from './db.js';

// Get local Cambodia (UTC+7) Date string and Time string
export const getLocalTimeDetails = (customTime, customDate) => {
  const now = new Date();
  const cambodiaTime = new Date(now.toLocaleString("en-US", { timeZone: "Asia/Phnom_Penh" }));

  let dateString;
  if (customDate) {
    dateString = customDate; // Format: "YYYY-MM-DD"
  } else {
    const year = cambodiaTime.getFullYear();
    const month = String(cambodiaTime.getMonth() + 1).padStart(2, '0');
    const day = String(cambodiaTime.getDate()).padStart(2, '0');
    dateString = `${year}-${month}-${day}`;
  }

  let timeString;
  if (customTime) {
    timeString = customTime; // Format: "HH:mm"
  } else {
    const hours = String(cambodiaTime.getHours()).padStart(2, '0');
    const minutes = String(cambodiaTime.getMinutes()).padStart(2, '0');
    timeString = `${hours}:${minutes}`;
  }

  return {
    dateObj: new Date(dateString),
    dateString,
    timeString
  };
};



// Convert "HH:mm" string to minutes from midnight
const timeToMinutes = (timeStr) => {
  if (!timeStr) return 0;
  const [h, m] = timeStr.split(':').map(Number);
  return h * 60 + m;
};

// Determine the action automatically based on current time and existing attendance
export const determineAutoAction = (employee, existingAttendance, timeString) => {
  const currentMinutes = timeToMinutes(timeString);
  const s1StartMinutes = timeToMinutes(employee?.shift1Start || '08:00');
  const s1EndMinutes = timeToMinutes(employee?.shift1End || '12:00');
  const s2StartMinutes = timeToMinutes(employee?.shift2Start || '13:00');
  const s2EndMinutes = timeToMinutes(employee?.shift2End || '17:00');

  const checkin1 = existingAttendance?.checkin1;
  const checkout1 = existingAttendance?.checkout1;
  const checkin2 = existingAttendance?.checkin2;
  const checkout2 = existingAttendance?.checkout2;

  const hasCheckIn1 = checkin1 && checkin1 !== '--:--' && checkin1 !== '-';
  const hasCheckOut1 = checkout1 && checkout1 !== '--:--' && checkout1 !== '-';
  const hasCheckIn2 = checkin2 && checkin2 !== '--:--' && checkin2 !== '-';
  const hasCheckOut2 = checkout2 && checkout2 !== '--:--' && checkout2 !== '-';

  if (hasCheckIn1 && hasCheckOut1 && hasCheckIn2 && hasCheckOut2) {
    return 'completed';
  }

  // 1. If scan time is past Shift 1 End (12:00 PM / afternoon)
  if (currentMinutes >= s1EndMinutes) {
    if (!hasCheckIn1) {
      // IF Scan_Time >= Shift 1 End and Check 1 is null -> return checkin_2
      if (!hasCheckIn2) return 'checkin_2';
      if (!hasCheckOut2) return 'checkout_2';
      return 'completed';
    } else {
      if (!hasCheckOut1 && currentMinutes <= s2StartMinutes) {
        return 'checkout_1';
      }
      if (!hasCheckIn2) return 'checkin_2';
      if (!hasCheckOut2) return 'checkout_2';
      return 'completed';
    }
  } else {
    // 2. Scan time is before Shift 1 End (morning)
    if (!hasCheckIn1) return 'checkin_1';
    if (!hasCheckOut1) return 'checkout_1';
    if (!hasCheckIn2) return 'checkin_2';
    if (!hasCheckOut2) return 'checkout_2';
    return 'completed';
  }
};

// Perform core attendance processing
export const processAttendanceScan = async ({ staffId, action: requestedAction, customTime, customDate, note }) => {
  // 1. Fetch employee
  const employee = await prisma.employee.findUnique({
    where: { staffId }
  });

  if (!employee) {
    throw new Error('Employee not found');
  }

  const { dateObj, timeString } = getLocalTimeDetails(customTime, customDate);

  // 2. Fetch existing attendance
  let attendance = await prisma.attendance.findUnique({
    where: {
      staffId_attendanceDate: {
        staffId,
        attendanceDate: dateObj
      }
    }
  });

  // 3. Determine action
  const action = requestedAction || determineAutoAction(employee, attendance, timeString);

  // 4. Update fields
  const updateData = {};
  if (note) updateData.note = note;

  if (action === 'checkin_1') {
    updateData.checkin1 = timeString;
  } else if (action === 'checkout_1') {
    updateData.checkout1 = timeString;
  } else if (action === 'checkin_2') {
    updateData.checkin2 = timeString;
  } else if (action === 'checkout_2') {
    updateData.checkout2 = timeString;
  }

  if (!attendance) {
    attendance = await prisma.attendance.create({
      data: {
        staffId,
        attendanceDate: dateObj,
        ...updateData
      }
    });
  } else {
    attendance = await prisma.attendance.update({
      where: { id: attendance.id },
      data: updateData
    });
  }

  // 5. Recalculate late and early metrics
  const c1 = attendance.checkin1 || updateData.checkin1;
  const c2 = attendance.checkin2 || updateData.checkin2;
  const o1 = attendance.checkout1 || updateData.checkout1;
  const o2 = attendance.checkout2 || updateData.checkout2;

  const s1Start = employee.shift1Start;
  const s1End = employee.shift1End;
  const s2Start = employee.shift2Start;
  const s2End = employee.shift2End;

  let isLate = false;
  let isEarlyLeave = false;

  if (c1 && c1 > s1Start) isLate = true;
  if (c2 && c2 > s2Start) isLate = true;
  if (o1 && o1 < s1End) isEarlyLeave = true;
  if (o2 && o2 < s2End) isEarlyLeave = true;

  const finalAttendance = await prisma.attendance.update({
    where: { id: attendance.id },
    data: {
      isLate,
      isEarlyLeave
    },
    include: {
      employee: {
        select: {
          staffId: true,
          nameEn: true,
          nameKh: true,
          branch: true,
          department: { select: { nameEn: true, nameKh: true } }
        }
      }
    }
  });

  return {
    attendance: finalAttendance,
    action,
    timeString,
    dateString: dateObj.toISOString().split('T')[0]
  };
};

File: backend/src/main/java/com/hrchomnan/backend/util/AttendanceHelper.java
package com.hrchomnan.backend.util;

import com.hrchomnan.backend.model.Attendance;
import com.hrchomnan.backend.model.CompanyWorkHour;
import com.hrchomnan.backend.model.Employee;
import com.hrchomnan.backend.repository.AttendanceRepository;
import com.hrchomnan.backend.repository.CompanyWorkHourRepository;
import com.hrchomnan.backend.repository.EmployeeRepository;
import com.hrchomnan.backend.service.TelegramNotificationService;
import lombok.Builder;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Optional;

@Slf4j
@Component
@Transactional
@RequiredArgsConstructor
public class AttendanceHelper {

    private static final ZoneId CAMBODIA_ZONE = ZoneId.of("Asia/Phnom_Penh");
    private final AttendanceRepository attendanceRepository;
    private final EmployeeRepository employeeRepository;
    private final CompanyWorkHourRepository companyWorkHourRepository;
    private final TelegramNotificationService telegramNotificationService;

    @Data
    @Builder
    public static class TimeDetails {
        private LocalDate date;
        private String dateString;
        private String timeString;
    }

    @Data
    @Builder
    public static class ScanResult {
        private Attendance attendance;
        private Employee employee;
        private String action;
        private String timeString;
        private String dateString;
    }

    public TimeDetails getLocalTimeDetails(String customTime, String customDate) {
        ZonedDateTime cambodiaNow = ZonedDateTime.now(CAMBODIA_ZONE);

        LocalDate date;
        if (customDate != null && !customDate.isBlank()) {
            date = LocalDate.parse(customDate.trim());
        } else {
            date = cambodiaNow.toLocalDate();
        }

        String timeString;
        if (customTime != null && !customTime.isBlank()) {
            timeString = customTime.trim();
        } else {
            timeString = cambodiaNow.toLocalTime().format(DateTimeFormatter.ofPattern("HH:mm"));
        }

        return TimeDetails.builder()
                .date(date)
                .dateString(date.toString())
                .timeString(timeString)
                .build();
    }

    public int timeToMinutes(String timeStr) {
        if (timeStr == null || !timeStr.contains(":")) return 0;
        try {
            String[] parts = timeStr.trim().split(":");
            int h = Integer.parseInt(parts[0]);
            int m = Integer.parseInt(parts[1]);
            return h * 60 + m;
        } catch (Exception e) {
            return 0;
        }
    }

    public String determineAutoAction(Employee employee, Attendance existingAttendance, String timeString) {
        int currentMinutes = timeToMinutes(timeString);
        int s1StartMinutes = timeToMinutes(employee != null && employee.getShift1Start() != null ? employee.getShift1Start() : "08:00");
        int s1EndMinutes = timeToMinutes(employee != null && employee.getShift1End() != null ? employee.getShift1End() : "12:00");
        int s2StartMinutes = timeToMinutes(employee != null && employee.getShift2Start() != null ? employee.getShift2Start() : "13:00");
        int s2EndMinutes = timeToMinutes(employee != null && employee.getShift2End() != null ? employee.getShift2End() : "17:00");

        String checkin1 = existingAttendance != null ? existingAttendance.getCheckin1() : null;
        String checkout1 = existingAttendance != null ? existingAttendance.getCheckout1() : null;
        String checkin2 = existingAttendance != null ? existingAttendance.getCheckin2() : null;
        String checkout2 = existingAttendance != null ? existingAttendance.getCheckout2() : null;

        boolean hasCheckIn1 = checkin1 != null && !checkin1.isBlank() && !checkin1.equals("--:--") && !checkin1.equals("-");
        boolean hasCheckOut1 = checkout1 != null && !checkout1.isBlank() && !checkout1.equals("--:--") && !checkout1.equals("-");
        boolean hasCheckIn2 = checkin2 != null && !checkin2.isBlank() && !checkin2.equals("--:--") && !checkin2.equals("-");
        boolean hasCheckOut2 = checkout2 != null && !checkout2.isBlank() && !checkout2.equals("--:--") && !checkout2.equals("-");

        // 1. IF (currentTime < s1_end):
        if (currentMinutes < s1EndMinutes) {
            if (!hasCheckIn1) return "checkin_1";
            else return "completed";
        }
        // 2. ELSE IF (currentTime >= s1EndMinutes AND currentTime < s2EndMinutes):
        else if (currentMinutes >= s1EndMinutes && currentMinutes < s2EndMinutes) {
            // ករណីភ្លេច Check-out វេនព្រឹក (ដល់/ហួសម៉ោង s1_end)
            if (hasCheckIn1 && !hasCheckOut1) return "checkout_1";
            // ករណីវេនព្រឹកចប់សព្វគ្រប់ ហើយចូលវេនរសៀល
            else if (!hasCheckIn2) return "checkin_2";
            else return "completed";
        }
        // 3. ELSE IF (currentTime >= s2_end):
        else {
            // ករណីដល់/ហួសម៉ោងវេនរសៀល
            if (hasCheckIn2 && !hasCheckOut2) return "checkout_2";
            else return "completed";
        }
    }

    public ScanResult processAttendanceScan(String staffId, String requestedAction, String customTime, String customDate, String note) {
        Optional<Employee> empOpt = employeeRepository.findByStaffId(staffId);
        if (empOpt.isEmpty()) {
            throw new RuntimeException("Employee not found");
        }
        Employee employee = empOpt.get();

        TimeDetails timeDetails = getLocalTimeDetails(customTime, customDate);
        LocalDate attendanceDate = timeDetails.getDate();
        String timeString = timeDetails.getTimeString();

        Optional<Attendance> existingOpt = attendanceRepository.findByStaffIdAndAttendanceDate(staffId, attendanceDate);
        boolean hadCheckout2Already = existingOpt.isPresent() &&
                existingOpt.get().getCheckout2() != null &&
                !existingOpt.get().getCheckout2().isBlank() &&
                !existingOpt.get().getCheckout2().equals("--:--") &&
                !existingOpt.get().getCheckout2().equals("-");

        Attendance attendance = existingOpt.orElseGet(() -> Attendance.builder()
                .staffId(staffId)
                .attendanceDate(attendanceDate)
                .build());

        String action;
        if (hadCheckout2Already) {
            action = "completed";
        } else {
            action = (requestedAction != null && !requestedAction.isBlank())
                    ? requestedAction
                    : determineAutoAction(employee, attendance, timeString);
        }

        if (note != null && !note.isBlank()) {
            attendance.setNote(note);
        }

        switch (action) {
            case "checkin_1" -> attendance.setCheckin1(timeString);
            case "checkout_1" -> attendance.setCheckout1(timeString);
            case "checkin_2" -> attendance.setCheckin2(timeString);
            case "checkout_2" -> attendance.setCheckout2(timeString);
        }

        // Recalculate late and early leave metrics
        String c1 = attendance.getCheckin1();
        String c2 = attendance.getCheckin2();
        String o1 = attendance.getCheckout1();
        String o2 = attendance.getCheckout2();

        String s1Start = (employee.getShift1Start() != null && !employee.getShift1Start().isBlank()) ? employee.getShift1Start() : "08:00";
        String s1End = (employee.getShift1End() != null && !employee.getShift1End().isBlank()) ? employee.getShift1End() : "12:00";
        String s2Start = (employee.getShift2Start() != null && !employee.getShift2Start().isBlank()) ? employee.getShift2Start() : "13:00";
        String s2End = (employee.getShift2End() != null && !employee.getShift2End().isBlank()) ? employee.getShift2End() : "17:00";

        int graceMinutes = 0;
        List<CompanyWorkHour> cwhList = companyWorkHourRepository.findAll();
        if (!cwhList.isEmpty() && cwhList.get(0).getLateGraceMinutes() != null) {
            graceMinutes = cwhList.get(0).getLateGraceMinutes();
        }

        int s1StartMin = timeToMinutes(s1Start) + graceMinutes;
        int s1EndMin = timeToMinutes(s1End);
        int s2StartMin = timeToMinutes(s2Start) + graceMinutes;
        int s2EndMin = timeToMinutes(s2End);

        boolean isLate = false;
        boolean isEarlyLeave = false;

        if (c1 != null && !c1.isBlank() && timeToMinutes(c1) > s1StartMin) isLate = true;
        if (c2 != null && !c2.isBlank() && timeToMinutes(c2) > s2StartMin) isLate = true;
        if (o1 != null && !o1.isBlank() && timeToMinutes(o1) < s1EndMin) isEarlyLeave = true;
        if (o2 != null && !o2.isBlank() && timeToMinutes(o2) < s2EndMin) isEarlyLeave = true;

        attendance.setIsLate(isLate);
        attendance.setIsEarlyLeave(isEarlyLeave);

        Attendance savedAttendance = attendanceRepository.save(attendance);

        // Send instant notification to Telegram Group ONLY if shift out 2 did not already have data
        if (!hadCheckout2Already && !"completed".equalsIgnoreCase(action)) {
            telegramNotificationService.sendAttendanceNotification(
                    employee,
                    savedAttendance,
                    action,
                    timeString,
                    null,
                    note
            );
        } else {
            log.info("Skipping Telegram notification for staffId {} - shift out 2 already has data for today or action is completed.", staffId);
        }

        return ScanResult.builder()
                .attendance(savedAttendance)
                .employee(employee)
                .action(action)
                .timeString(timeString)
                .dateString(timeDetails.getDateString())
                .build();
    }
}
